language/vm/vm_runtime/src/interpreter.rs

```diff
@@ -0,0 +1,1204 @@
+    txn_data: TransactionMetadata,
+    /// List of events "fired" during the course of an execution.
+    // REVIEW: should this live outside the Interpreter?
+    event_data: Vec<ContractEvent>,
+    /// Data store
+    // REVIEW: maybe this and the event should go together as some kind of external context?
+    data_view: TransactionDataCache<'txn>,
+    /// Code cache, this is effectively the loader.
+    module_cache: P,
+    phantom: PhantomData<&'alloc ()>,
+}
+
+impl<'alloc, 'txn, P> Interpreter<'alloc, 'txn, P>
+where
+    'alloc: 'txn,
+    P: ModuleCache<'alloc>,
+{
+    /// Create a new instance of an `Interpreter` in the context of a transaction with a
+    /// given module cache (loader) and a data store.
+    // REVIEW: it's not clear the responsibilities between the Interpreter and the outside
+    // context are well defined. Obviously certain opcodes require a given context, but
+    // we may be doing better here...
+    pub fn new(
+        module_cache: P,
+        txn_data: TransactionMetadata,
+        data_view: TransactionDataCache<'txn>,
+    ) -> Self {
+        Interpreter {
+            operand_stack: Stack::new(),
+            call_stack: CallStack::new(),
+            gas_meter: GasMeter::new(txn_data.max_gas_amount()),
+            txn_data,
+            event_data: vec![],
+            data_view,
+            module_cache,
+            phantom: PhantomData,
+        }
+    }
+
+    //
+    // The functions below should be reviewed once we clean up the Loader and the
+    // transaction flow. It's not clear whether they are leaking internal of the Interpreter
+    // that would be better exposed via a more proper API.
+    //
+
+    /// Returns the module cache for this interpreter.
+    pub fn module_cache(&self) -> &P {
+        &self.module_cache
+    }
+
+    /// Disables metering of gas.
+    pub fn disable_metering(&mut self) {
+        self.gas_meter.disable_metering();
+    }
+
+    /// Re-enables metering of gas.
+    pub(crate) fn enable_metering(&mut self) {
+        self.gas_meter.enable_metering();
+    }
+
+    /// Returns the gas used by an execution in the `Interpreter`.
+    pub(crate) fn gas_used(&self) -> u64 {
+        self.txn_data
+            .max_gas_amount
+            .sub(self.gas_meter.remaining_gas())
+            .mul(self.txn_data.gas_unit_price)
+            .get()
+    }
+
+    // This is used by the genesis tool and must be deleted and re-worked
+    pub(crate) fn swap_sender(&mut self, address: AccountAddress) -> AccountAddress {
+        let old_sender = self.txn_data.sender;
+        self.txn_data.sender = address;
+        old_sender
+    }
+
+    /// Clear all the writes local to this execution.
+    pub(crate) fn clear(&mut self) {
+        self.data_view.clear();
+        self.event_data.clear();
+    }
+
+    /// Return the list of events emitted during execution.
+    pub(crate) fn get_events(&self) -> Vec<ContractEvent> {
+        self.event_data.clone()
+    }
```

Comment:
Generally better to let the caller clone or use `to_vec` if necessary. Also `get_` is typically not used for a simple getter.

```
    pub(crate) fn events(&self) -> &[ContractEvent] {
        &self.event_data
    }
```

Reply:
done